crates/bevy_pbr/src/render/gpu_preprocess.rs

```diff
@@ -0,0 +1,305 @@
+                    )
+                    .in_set(RenderSet::PrepareBindGroups),
+            ),
+        );
+    }
+
+    fn finish(&self, app: &mut App) {
+        let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        // Stitch the node in.
+        render_app
+            .add_render_graph_node::<GpuPreprocessNode>(Core3d, NodePbr::GpuPreprocess)
+            .add_render_graph_edges(Core3d, (NodePbr::GpuPreprocess, Node3d::Prepass))
+            .init_resource::<PreprocessPipeline>()
+            .init_resource::<SpecializedComputePipelines<PreprocessPipeline>>();
+    }
+}
+
+impl FromWorld for GpuPreprocessNode {
+    fn from_world(world: &mut World) -> Self {
+        Self {
+            view_query: QueryState::new(world),
+        }
+    }
+}
+
+impl Node for GpuPreprocessNode {
+    fn update(&mut self, world: &mut World) {
+        self.view_query.update_archetypes(world);
+    }
+
+    fn run<'w>(
+        &self,
+        _: &mut RenderGraphContext,
+        render_context: &mut RenderContext<'w>,
+        world: &'w World,
+    ) -> Result<(), NodeRunError> {
+        // Grab the [`BatchedInstanceBuffers`]. If we aren't using GPU mesh
+        // uniform building, bail out.
+        let Some(BatchedInstanceBuffers {
+            work_item_buffers: ref index_buffers,
+            ..
+        }) = world.get_resource::<BatchedInstanceBuffers<MeshUniform, MeshInputUniform>>()
+        else {
+            error!(
+                "Attempted to preprocess meshes on GPU, but `GpuBuilt` batched instance buffers \
+                weren't available"
+            );
+            return Ok(());
+        };
+
+        let pipeline_cache = world.resource::<PipelineCache>();
+        let preprocess_pipeline = world.resource::<PreprocessPipeline>();
+
+        let Some(preprocess_pipeline_id) = preprocess_pipeline.pipeline_id else {
+            warn!("The build mesh uniforms pipeline wasn't uploaded");
+            return Ok(());
+        };
+
+        let Some(preprocess_pipeline) = pipeline_cache.get_compute_pipeline(preprocess_pipeline_id)
+        else {
+            // This will happen while the pipeline is being compiled and is fine.
+            return Ok(());
+        };
+
+        let mut compute_pass =
+            render_context
+                .command_encoder()
+                .begin_compute_pass(&ComputePassDescriptor {
+                    label: Some("mesh preprocessing"),
+                    timestamp_writes: None,
+                });
+
+        compute_pass.set_pipeline(preprocess_pipeline);
+
+        for (view, bind_group) in self.view_query.iter_manual(world) {
+            // Grab the index buffer for this view.
+            let Some(index_buffer) = index_buffers.get(&view) else {
+                warn!("The preprocessing index buffer wasn't present");
+                return Ok(());
```

Comment:
Return? Or continue?

Reply:
And how do we prevent passes running that depend on this pass having been run successfully?

Reply:
Good point, I added a check in `DrawMesh`. I couldn't add it anywhere else because that check needs to go in `bevy_pbr` and not any upstream crate. Thankfully the only draw command that makes use of the mesh instance buffer is `DrawMesh`, so I only had to update one place.